Add reference target

dandclark opened the pull request:
Reference Target allows authors to specify an element inside a shadow root to be the target of any ID references referring to the host element. This would enable IDREF attributes such as for and aria-labelledby to refer to elements inside a component's shadow DOM while maintaining encapsulation of the internal details of the shadow DOM.
 
See the [reference target explainer](https://github.com/WICG/webcomponents/blob/gh-pages/proposals/reference-target-explainer.md).

At a high level, the spec change consists of these parts:
- Define the concept of "resolving the reference target" on an element, following an IDREF from a host element to its target in the shadow, and potentially recursing.
- Update the `get-the-attr-associated-element` algorithm (and its array-of-elements form) to follow reference target.
- Centralize the definitions of ID reference attributes and list-of-ID-reference attributes as "element reference" and "set of element reference" attributes instead of repeating these definitions throughout the spec.
- Update uses of ID throughout the spec, where applicable, to use the new reference-target-aware `get-the-attr-associated-element`.
- Avoid exposing reference target elements in shadows from element-returning IDL properties by [retargeting](https://dom.spec.whatwg.org/#retarget) the result before returning.

See also the corresponding https://github.com/whatwg/dom/pull/1353 which adds the definition of reference target used in this PR.
 
<!--
Thank you for contributing to the HTML Standard! Please describe the change you are making and complete the checklist below if your change is not editorial.

When you submit this PR, and each time you edit this comment (including checking a checkbox through the UI!), PR Preview will run and update it. As such make any edits in one go and only after PR Preview has run.

If you think your PR is ready to land, please double-check that the build is passing and the checklist is complete before pinging.
-->

- [X] At least two implementers are interested (and none opposed):
   * Chromium is positive (I'd like to ship it there)
   * https://github.com/mozilla/standards-positions/issues/1035 (Positive)
   * https://github.com/WebKit/standards-positions/issues/356 (no position stated yet)
- [X] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * https://wpt.fyi/results/shadow-dom/reference-target/tentative?label=experimental&label=master&aligned (still tentative)
   * PR to make non-tentative: https://github.com/web-platform-tests/wpt/pull/51213
- [X] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chromium: https://issues.chromium.org/issues/346835896/
   * Gecko: https://bugzilla.mozilla.org/show_bug.cgi?id=1952585
   * WebKit: https://bugs.webkit.org/show_bug.cgi?id=285634
- [X] Corresponding [HTML AAM](https://w3c.github.io/html-aam/) & [ARIA in HTML](https://w3c.github.io/html-aria/) issues & PRs: https://github.com/w3c/aria/pull/2474
- [X] [MDN issue](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) is filed: https://github.com/mdn/mdn/issues/832
- [x] The top of this comment includes a [clear commit message](https://github.com/whatwg/meta/blob/main/COMMITTING.md) to use. <!-- If you created this PR from a single commit, Github copied its message. Otherwise, you need to add a commit message yourself. -->

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/10995/common-dom-interfaces.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">/common-dom-interfaces.html</a>  ( <a href="https://whatpr.org/html/10995/320c05f...0818c8e/common-dom-interfaces.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">diff</a> )
<a href="https://whatpr.org/html/10995/common-microsyntaxes.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">/common-microsyntaxes.html</a>  ( <a href="https://whatpr.org/html/10995/320c05f...0818c8e/common-microsyntaxes.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">diff</a> )
<a href="https://whatpr.org/html/10995/document-lifecycle.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">/document-lifecycle.html</a>  ( <a href="https://whatpr.org/html/10995/320c05f...0818c8e/document-lifecycle.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">diff</a> )
<a href="https://whatpr.org/html/10995/dom.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">/dom.html</a>  ( <a href="https://whatpr.org/html/10995/320c05f...0818c8e/dom.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">diff</a> )
<a href="https://whatpr.org/html/10995/form-control-infrastructure.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">/form-control-infrastructure.html</a>  ( <a href="https://whatpr.org/html/10995/320c05f...0818c8e/form-control-infrastructure.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">diff</a> )
<a href="https://whatpr.org/html/10995/form-elements.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">/form-elements.html</a>  ( <a href="https://whatpr.org/html/10995/320c05f...0818c8e/form-elements.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">diff</a> )
<a href="https://whatpr.org/html/10995/forms.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">/forms.html</a>  ( <a href="https://whatpr.org/html/10995/320c05f...0818c8e/forms.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">diff</a> )
<a href="https://whatpr.org/html/10995/iframe-embed-object.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">/iframe-embed-object.html</a>  ( <a href="https://whatpr.org/html/10995/320c05f...0818c8e/iframe-embed-object.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">diff</a> )
<a href="https://whatpr.org/html/10995/index.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">/index.html</a>  ( <a href="https://whatpr.org/html/10995/320c05f...0818c8e/index.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">diff</a> )
<a href="https://whatpr.org/html/10995/indices.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">/indices.html</a>  ( <a href="https://whatpr.org/html/10995/320c05f...0818c8e/indices.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">diff</a> )
<a href="https://whatpr.org/html/10995/infrastructure.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">/infrastructure.html</a>  ( <a href="https://whatpr.org/html/10995/320c05f...0818c8e/infrastructure.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">diff</a> )
<a href="https://whatpr.org/html/10995/input.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">/input.html</a>  ( <a href="https://whatpr.org/html/10995/320c05f...0818c8e/input.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">diff</a> )
<a href="https://whatpr.org/html/10995/interaction.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">/interaction.html</a>  ( <a href="https://whatpr.org/html/10995/320c05f...0818c8e/interaction.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">diff</a> )
<a href="https://whatpr.org/html/10995/interactive-elements.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">/interactive-elements.html</a>  ( <a href="https://whatpr.org/html/10995/320c05f...0818c8e/interactive-elements.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">diff</a> )
<a href="https://whatpr.org/html/10995/links.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">/links.html</a>  ( <a href="https://whatpr.org/html/10995/320c05f...0818c8e/links.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">diff</a> )
<a href="https://whatpr.org/html/10995/microdata.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">/microdata.html</a>  ( <a href="https://whatpr.org/html/10995/320c05f...0818c8e/microdata.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">diff</a> )
<a href="https://whatpr.org/html/10995/nav-history-apis.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">/nav-history-apis.html</a>  ( <a href="https://whatpr.org/html/10995/320c05f...0818c8e/nav-history-apis.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">diff</a> )
<a href="https://whatpr.org/html/10995/parsing.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">/parsing.html</a>  ( <a href="https://whatpr.org/html/10995/320c05f...0818c8e/parsing.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">diff</a> )
<a href="https://whatpr.org/html/10995/popover.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">/popover.html</a>  ( <a href="https://whatpr.org/html/10995/320c05f...0818c8e/popover.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">diff</a> )
<a href="https://whatpr.org/html/10995/scripting.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">/scripting.html</a>  ( <a href="https://whatpr.org/html/10995/320c05f...0818c8e/scripting.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">diff</a> )
<a href="https://whatpr.org/html/10995/tables.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">/tables.html</a>  ( <a href="https://whatpr.org/html/10995/320c05f...0818c8e/tables.html" title="Last updated on Jul 29, 2026, 4:09 PM UTC (0818c8e)">diff</a> )

alice committed:
Begin exploring how to specify what effect referenceTarget has.

alice committed:
Begin exploring how to specify what effect referenceTarget has.

dandclark committed:
Add shadowrootreferencetarget attribute to template, plumb it through to shadow creation

dandclark committed:
Add shadowrootreferencetarget attribute to template, plumb it through to shadow creation

dandclark committed:
Use 'reference target' internal definition on Shadow instead of IDL property

dandclark committed:
Use 'reference target' internal definition on Shadow instead of IDL property

dandclark committed:
Implicit label association accounts for reference target

dandclark committed:
Implicit label association accounts for reference target

dandclark committed:
Don't directly nest <ol>s

dandclark committed:
Don't directly nest <ol>s

dandclark committed:
Actually return the resolved reference target from labeled control algo

dandclark committed:
Actually return the resolved reference target from labeled control algo

alice committed:
Add a section about multiple element reference attributes

alice committed:
Add a section about multiple element reference attributes

dandclark committed:
Update HTMLLabelElement control and form properties

dandclark committed:
Update HTMLLabelElement control and form properties

dandclark committed:
Update the HTMLInputElement.list attribute

dandclark committed:
Update the HTMLInputElement.list attribute

dandclark committed:
Update headers property

dandclark committed:
Update headers property

dandclark committed:
Update form association

dandclark committed:
Update form association

alice committed:
Define headers and list attributes in terms of element reference attribute concepts.

alice committed:
Define headers and list attributes in terms of element reference attribute concepts.

alice committed:
Update output-for and popover, and reword definitions

alice committed:
Update output-for and popover, and reword definitions

dandclark committed:
Update itemref attribute

dandclark committed:
Update itemref attribute

dandclark committed:
Update attributes table

dandclark committed:
Update attributes table

alice committed:
Add notes an examples in the Element reference attributes section, and make some tweaks in the Reflection section

alice committed:
Add notes an examples in the Element reference attributes section, and make some tweaks in the Reflection section

alice committed:
A few more minor tweaks

alice committed:
A few more minor tweaks

dandclark commented:
@annevk, @domenic, this isn't quite ready to land yet as there are still a few open questions to sort out (at least https://github.com/WICG/webcomponents/issues/1087 and https://github.com/WICG/webcomponents/issues/1093), plus missing implementor positions. But I don't expect most of the mechanics in this PR to change, so I think it would be great to get feedback on whether the approach we've used here is agreeable.

See also https://github.com/whatwg/dom/pull/1353

Thanks!

alice commented:
There's also a WIP PR against ARIA to account for these changes: https://github.com/w3c/aria/pull/2474

There are some open questions being discussed there on exactly how to refer to the HTML concepts, since ARIA needs to be language-independent, but it will certainly refer to these spec concepts in some form.

alice committed:
Add the concept of an event source.

This concept is used to propagate events into the source's tree under
certain circumstances.

lukewarlow commented on source:
```diff
@@ -64744,6 +64893,7 @@ interface <dfn interface>HTMLTemplateElement</dfn> : <span>HTMLElement</span> {
   [<span>CEReactions</span>] attribute boolean <span data-x="dom-template-shadowrootdelegatesfocus">shadowRootDelegatesFocus</span>;
   [<span>CEReactions</span>] attribute boolean <span data-x="dom-template-shadowrootclonable">shadowRootClonable</span>;
   [<span>CEReactions</span>] attribute boolean <span data-x="dom-template-shadowRootSerializable">shadowRootSerializable</span>;
+  [<span>CEReactions</span>] attribute DOMString? <span data-x="dom-template-shadowrootreferencetarget">shadowRootReferenceTarget</span>;
```
```suggestion
  [<span>CEReactions</span>, <span data-x="xattr-Reflect">Reflect</span>] attribute DOMString? <dfn attribute for="HTMLTemplateElement" data-x="dom-template-shadowrootreferencetarget">shadowRootReferenceTarget</dfn>;
```

lukewarlow commented on source:
```diff
@@ -64944,6 +65099,11 @@ interface <dfn interface>HTMLTemplateElement</dfn> : <span>HTMLElement</span> {
+  data-x="dom-template-shadowrootreferencetarget">shadowRootReferenceTarget</code></dfn> IDL
+  attribute must <span>reflect</span> the <code data-x="attr-template-shadowrootreferencetarget">
+  shadowrootreferencetarget</code> content attribute.</p>
+
```
```suggestion
```

lukewarlow commented on source:
```diff
@@ -64944,6 +65099,11 @@ interface <dfn interface>HTMLTemplateElement</dfn> : <span>HTMLElement</span> {
+  <p>The <dfn attribute for="HTMLTemplateElement"><code
+  data-x="dom-template-shadowrootreferencetarget">shadowRootReferenceTarget</code></dfn> IDL
+  attribute must <span>reflect</span> the <code data-x="attr-template-shadowrootreferencetarget">
+  shadowrootreferencetarget</code> content attribute.</p>
```
```suggestion
```

lukewarlow commented on source:
```diff
@@ -64944,6 +65099,11 @@ interface <dfn interface>HTMLTemplateElement</dfn> : <span>HTMLElement</span> {
 
+  <p>The <dfn attribute for="HTMLTemplateElement"><code
+  data-x="dom-template-shadowrootreferencetarget">shadowRootReferenceTarget</code></dfn> IDL
+  attribute must <span>reflect</span> the <code data-x="attr-template-shadowrootreferencetarget">
```
```suggestion
```

lukewarlow commented on source:
```diff
@@ -64944,6 +65099,11 @@ interface <dfn interface>HTMLTemplateElement</dfn> : <span>HTMLElement</span> {
   data-x="attr-template-shadowrootserializable">shadowrootserializable</code> content attribute.</p>
 
+  <p>The <dfn attribute for="HTMLTemplateElement"><code
+  data-x="dom-template-shadowrootreferencetarget">shadowRootReferenceTarget</code></dfn> IDL
```
```suggestion
```

lukewarlow commented on source:
```diff
@@ -64944,6 +65099,11 @@ interface <dfn interface>HTMLTemplateElement</dfn> : <span>HTMLElement</span> {
   must <span>reflect</span> the <code
   data-x="attr-template-shadowrootserializable">shadowrootserializable</code> content attribute.</p>
 
+  <p>The <dfn attribute for="HTMLTemplateElement"><code
```
```suggestion
```

lukewarlow reviewed (changes requested):
This should be updated to make use of the new [Reflect] syntax.

dandclark commented:
> This should be updated to make use of the new [Reflect] syntax.

Thanks @lukewarlow! Fixed now (it wouldn't let me merge the suggestions directly due to merge conflicts).

lukewarlow commented:
I'm not sure how GitHub actually lets you dismiss change suggestions but consider this me dropping my change suggestions as they've been resolved :)

keithamus commented on source:
```diff
@@ -5206,6 +5206,239 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+   <li><p>Let <var>referenceTarget</var> be the value of <var>element</var>'s <span>shadow
+   root</span>'s <span>reference target</span>.</p></li>
+
+   <li><p>Let <var>candidate</var> be the first element in <var>element</var>'s <span>shadow
+   root</span> whose <span data-x="concept-id">ID</span> matches <var>referenceTarget</var>.</p>
+   </li>
+
+   <li><p>If no such element exists, return null.</p></li>
```
These steps being disjoint were confusing to me, as step 4 read in isolation makes no reference to variables or some such. I think they'd be better joined:


```suggestion
   <li><p>Let <var>candidate</var> be the first element in <var>element</var>'s <span>shadow
   root</span> whose <span data-x="concept-id">ID</span> matches <var>referenceTarget</var>. If no
   such element exists, then return null.</p></li>
```

Or alternatively:


```suggestion
   <li><p>Let <var>candidate</var> be the first element in <var>element</var>'s <span>shadow
   root</span> whose <span data-x="concept-id">ID</span> matches <var>referenceTarget</var>. If no
   such element exists, then instead let <var>element</var> be null.</p></li>

   <li><p>If <var>element</var> is null, then return null</p></li>

```

dandclark replied:
Applied the second suggestion (with *candidate* instead of *element*).

keithamus commented on source:
```diff
@@ -5206,6 +5206,239 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+
+   <li><p>If no such element exists, return null.</p></li>
+
+   <li><p>Return the result of <span data-x="resolve-the-reference-target">resolving the reference
+   target</span> on <var>candidate</var>.
+   </p></li>
```
Formatting:

```suggestion
   <li><p>Return the result of <span data-x="resolve-the-reference-target">resolving the reference
   target</span> on <var>candidate</var>.</p></li>
```

keithamus commented on source:
```diff
@@ -5206,6 +5206,239 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+     <li><p>Otherwise, let <var>candidate</var> be null.</p></li>
+    </ol>
+
+    <p class="note">An element will only have an
+      <span data-x="explicitly set attr-element">explicitly-set <var>attr</var>-element</span> if it has
+      an IDL attribute with a type derived from <code>Element</code> which
+      <span data-x="reflect">reflects</span> <var>attr</var>.</p>
+   </li>
```
Formatting:


```suggestion
    <p class="note">An element will only have an
    <span data-x="explicitly set attr-element">explicitly-set <var>attr</var>-element</span> if it
    has an IDL attribute with a type derived from <code>Element</code> which
    <span data-x="reflect">reflects</span> <var>attr</var>.</p>
```

keithamus commented on source:
```diff
@@ -5206,6 +5206,239 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+   </li>
+
+   <li>
+    <p>Otherwise:
```
```suggestion
    <p>Otherwise:</p>
```

keithamus commented on source:
```diff
@@ -5206,6 +5206,239 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+
+  <h5>Sets of element references</h5>
+
+  <p>A <dfn data-x="set of element references"
+  data-lt="set of element references" export>set of element references</dfn> attribute represents
+  a set of references to elements. When specified as a content attribute, a <span>set of element
+  references</span> consists of a <span>set of space-separated tokens</span>, each of which should
+  completely match the <span data-x="concept-ID">ID</span> of another element in the document.</p>
```
Formatting:

```suggestion
  <p>A <dfn data-x="set of element references" data-lt="set of element references" export>set of
  element references</dfn> attribute represents a set of references to elements. When specified as
  a content attribute, a <span>set of element references</span> consists of a
  <span>set of space-separated tokens</span>, each of which should completely match the
  <span data-x="concept-ID">ID</span> of another element in the document.</p>
```

keithamus commented on source:
```diff
@@ -5206,6 +5206,239 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+      </ol>
+     </li>
+    </ol>
+
```
Double line break:

```suggestion
```

keithamus commented on source:
```diff
@@ -5206,6 +5206,239 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+    </ol>
+
+
+    <p class="note">An element will only have
+    <span data-x="explicitly set attr-elements">explicitly-set <var>attr</var>-elements</span> if it
+    has an IDL attribute with type <code
+    data-x="">FrozenArray&lt;<var>T</var>&gt;?</code>, where <var>T</var> is either
+    <code>Element</code> or an interface that inherits from <code>Element</code>, which
+    <span data-x="reflect">reflects</span> <var>attr</var>.</p>
```
Formatting seems wonky here, maybe this is better:


```suggestion
    <p class="note">An element will only have
    <span data-x="explicitly set attr-elements">explicitly-set <var>attr</var>-elements</span> if it
    has an IDL attribute with type <code data-x="">FrozenArray&lt;<var>T</var>&gt;?</code>, where
    <var>T</var> is either <code>Element</code> or an interface that inherits from
    <code>Element</code>, which <span data-x="reflect">reflects</span> <var>attr</var>.</p>
```

keithamus commented on source:
```diff
@@ -5206,6 +5206,239 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+
+   <li><p>Let <var>resolvedCandidates</var> be an empty list.</p></li>
+
+   <li>
+    <p><span data-x="list iterate">For each</span> <var>candidate</var> in <var>candidates</var>:
+    </p>
+
+    <ol>
+     <li><p>Let <var>resolvedCandidate</var> be the result of
+     <span data-x="resolve-the-reference-target">resolving the reference target</span> on
+     <var>candidate</var>.</p></li>
+
+     <li><p>If <var>resolvedCandidate</var> is not null, append <var>resolvedCandidate</var> to
+     <var>resolvedCandidates</var>.</p></li>
+    </ol>
+   </li>
+
+   <li><p>Return <var>resolvedCandidates</var>.</p></li>
+  </ol>
```
Why do we make a list of candidates to then re-iterate the list and resolve them? Would it not be easier to resolve each candidate during the for each steps so that `candidates` is a list of `resolvedCandidate`s?

keithamus commented on source:
```diff
@@ -8516,66 +8749,28 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
-      </ul>
-
-      <p>If no such element exists, then return null.</p>
+     <li><p>Let <var>candidate</var> be the result of running <span>this</span>'s
+     <span data-x="get the attr-associated element">get the <var>attr</var>-associated
+     element</span>.</p>
```
Oh I see. IDL reflections take the attr-resolved reference target and then retarget against element. This seems a little like extra work for no payoff. Are there concrete reasons to do this, rather than having attr-resolved elements return the element, and then in each spec resolve the reference target?

I'm also not sure retargeting against element is always going to consistently give the same result as the old way of handling things. This feels awkward to me.

keithamus commented on source:
```diff
@@ -46389,7 +46513,7 @@ interface <dfn interface>HTMLFormElement</dfn> : <span>HTMLElement</span> {
 interface <dfn interface>HTMLLabelElement</dfn> : <span>HTMLElement</span> {
   [<span>HTMLConstructor</span>] constructor();
 
-  readonly attribute <span>HTMLFormElement</span>? <span data-x="dom-label-form">form</span>;
+  readonly attribute <span>HTMLElement</span>? <span data-x="dom-label-form">form</span>;
```
Do we know if changing this type is web compatible? I understand the reasons for it but I feel a little hesitant.

keithamus reviewed (changes requested):
I've done an initial review pass, got about half way, but I thought I'd provide my review commentary so far because I think if I continue I may end up repeating themes. 

Some meta commentary:

 - I think another formatting pass needs to be done.
 - I am a little worried about the retargeting steps, I wonder if this can be simplified to avoid retargeting and to resolve a reference target where necessary.

smaug---- commented on source:
```diff
@@ -57070,8 +57220,9 @@ interface <dfn interface>HTMLSelectedContentElement</dfn> : <span>HTMLElement</s
-  <span>tree</span>.</p>
+  a valid <span>single element reference</span> attribute
+  <span data-x="single-element-reference-refers-to">referring to</span> a <code>form</code>
+  element.</p>
```
This isn't perhaps super clear. I _think_ this means that the final element from `single element reference` should be a form element

lukewarlow commented on source:
```diff
@@ -5206,6 +5206,239 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+    <ol>
+     <li>
+      <p><span data-x="list iterate">For each</span> <var>attrElement</var> in
+      <var>reflectedTarget</var>'s <span>explicitly set <var>attr</var>-elements</span>:</p>
```
```suggestion
      <var>element</var>'s <span>explicitly set <var>attr</var>-elements</span>:</p>
```

lukewarlow commented on source:
```diff
@@ -5206,6 +5206,239 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+   <li><p>Let <var>candidates</var> be an empty <span>list</span>.</p></li>
+
+   <li>
+    <p>If <var>element</var>'s has an <span>explicitly set <var>attr</var>-elements</span> which
```
This isn't grammatically correct. Based on the which I'm guessing the has is correct so drop the 's
```suggestion
    <p>If <var>element</var> has an <span>explicitly set <var>attr</var>-elements</span> which
```

smaug---- commented on source:
```diff
@@ -46399,33 +46523,50 @@ interface <dfn interface>HTMLLabelElement</dfn> : <span>HTMLElement</span> {
-  order</span> is a <span data-x="category-label">labelable element</span>, then that element is the
-  <code>label</code> element's <span>labeled control</span>.</span></p>
+  attribute is specified, the attribute's value must be the a valid <span>single element
+  reference</span> attribute.</p>
```
But the final result of `single element reference` doesn't need to be labelable element?

smaug---- commented on source:
```diff
@@ -88086,9 +88236,8 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
-  data-x="attr-popover">popover</code> attribute in the same <span>tree</span> as the <span
-  data-x="concept-button">button</span> with the <code
+  be a valid <span>single element reference</span> attribute
+  <span data-x="single-element-reference-refers-to">referring to</span> an element with the <code
```
Is this correct? The old text requires an element with popover attribute.

alice replied:
Huh, yeah, something went weird there. Fixed.

smaug---- commented on source:
```diff
@@ -65213,6 +65362,7 @@ interface <dfn interface>HTMLTemplateElement</dfn> : <span>HTMLElement</span> {
   [<span>CEReactions</span>, <span data-x="xattr-Reflect">Reflect</span>] attribute boolean <dfn attribute for="HTMLTemplateElement" data-x="dom-template-shadowrootclonable">shadowRootClonable</dfn>;
   [<span>CEReactions</span>, <span data-x="xattr-Reflect">Reflect</span>] attribute boolean <dfn attribute for="HTMLTemplateElement" data-x="dom-template-shadowRootSerializable">shadowRootSerializable</dfn>;
   [<span>CEReactions</span>, <span data-x="xattr-Reflect">Reflect</span>] attribute DOMString <dfn attribute for="HTMLTemplateElement" data-x="dom-template-shadowRootCustomElementRegistry">shadowRootCustomElementRegistry</dfn>;
+  [<span>CEReactions</span>, <span data-x="xattr-Reflect">Reflect</span>] attribute DOMString? <dfn attribute for="HTMLTemplateElement" data-x="dom-template-shadowRootReferenceTarget">shadowRootReferenceTarget</dfn>;
```
Why is shadowRootReferenceTarget nullable?

alice replied:
This is an open issue: https://github.com/WICG/webcomponents/issues/1093

alice reviewed (commented):
I had to apply all the code suggestions manually, because I foolishly pushed my changes before I'd applied them.

alice commented on source:
```diff
@@ -5206,6 +5206,235 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+    </ol>
+   </li>
+
+   <li>If candidate is null, return null.</li>
```
```suggestion
   <li><p>If candidate is null, return null.</p></li>
```

alice committed:
Address Anne's review comments

alice committed:
Address Anne's review comments

keithamus reviewed (commented):
Couple small issues:

alice committed:
Split attr-associated elements into two different algorithms: resolved attr target element(s) and unresolved attr target element(s)

alice committed:
Split attr-associated elements into two different algorithms: resolved attr target element(s) and unresolved attr target element(s)

dandclark committed:
Define source on command event

dandclark committed:
Fix a couple bits of feedback that were missed and clean up some formatting.

annevk commented on source:
```diff
@@ -5325,6 +5333,274 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+  <p><dfn data-x="element reference attribute" data-lt="element reference" export>Element reference
+  attributes</dfn> refer to one or more elements in the document as their target(s). When specified
+  as content attributes, <span data-x="element reference attribute">element reference
+  attributes</span> refer to other elements via their <span data-x="concept-ID">ID</span>s.</p>
```
```suggestion
  attributes</span> refer to other elements via their <span data-x="concept-ID">IDs</span>.</p>
```

dandclark replied:
For some reason I can never get suggestions on HTML spec PRs to commit directly. There's always an error that I need to refresh the page, which doesn't go away on refresh.

Anyway, pushed this change.

annevk commented on source:
```diff
@@ -5325,6 +5333,274 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+
+   <li><p>Let <var>candidate</var> be the first element in <var>element</var>'s <span>shadow
+   root</span> whose <span data-x="concept-id">ID</span> matches <var>referenceTargetValue</var>. If
+   no such element exists, then instead let <var>candidate</var> be null.</p></li>
```
We don't want nested assignment like this. Either initialize using a ternary (see Infra) or do it with two steps (and use set in the second step).

dandclark replied:
Broke into two steps.

annevk commented on source:
```diff
@@ -5325,6 +5333,274 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+
+  <h5>Single-element references</h5>
+
+  <p>A <dfn data-x="single-element reference" data-lt="single-element reference"
```
I'm not sure why we have data-lt or data-x here.

dandclark replied:
Removed both.

annevk commented on source:
```diff
@@ -5325,6 +5333,274 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+
+  <p>A <dfn data-x="single-element reference" data-lt="single-element reference"
+  export>single-element reference</dfn> attribute represents a reference to a single element.
+  When specified as a content attribute, a <span>single-element reference</span> consists of a
```
As a content attribute value? Maybe even an element-reference attribute value?

dandclark replied:
Changed to "When specified as a content attribute value..."

annevk commented on source:
```diff
@@ -5325,6 +5333,274 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
 
+  <h4>Element reference attributes</h4>
+
+  <p><dfn data-x="element reference attribute" data-lt="element reference" export>Element reference
```
I don't understand why we have data-x or data-lt here. Why is this one not exported but the value type is?

I think it should also have a hyphen (element-reference attribute).

dandclark replied:
Removed data-x and data-lt.

Both definitions are exported -- both likely need to be referenced from ARIA: https://github.com/w3c/aria/pull/2474

Added the hyphen. This looks a little inconsistent now though compared to "single-element reference" and "set of element reference", which has no hyphen between "element" and "reference". But "single-element-reference" would look strange.

annevk commented on source:
```diff
@@ -5325,6 +5333,274 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+
+  <div algorithm>
+  <p>A <span>single-element reference</span> attribute <var>attr</var> on an element <var>X</var>
+  <dfn data-x="single-element-reference-refers-to">refers to</dfn> an element <var>Y</var> as its
```
Instead of _attr_ we should do _attribute_. Instead of X and Y I think reference and target would fit?

dandclark replied:
Changed to `attribute`.

I think "reference" could be a confusing name for the element doing the referencing. The "reference" is more the attribute itself, while the element is the "referencer", if that was a word.

How about "source" and "target"? I do like the use of "target" because then we can drop the "as its target", and it just becomes:

```html
  <p>A <span>single-element reference</span> attribute <var>attribute</var> on an element
  <var>source</var> <dfn data-x="single-element-reference-refers-to">refers to</dfn> an element
  <var>target</var> if <var>target</var> is the <span data-x="get the resolved attribute target
  element">resolved <var>attribute</var> target element</span> for <var>source</var>.</p>
```

annevk replied:
Seems fine.

annevk commented on source:
```diff
@@ -5325,6 +5333,274 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+   <span data-x="get the unresolved attr target element">unresolved <var>attr</var> target
+   element</span>.</p></li>
+
+   <li><p>If <var>unresolvedTarget</var> is null, return null.</p></li>
```
then return* (this prolly applies elsewhere, please check carefully)

dandclark replied:
Fixed here and the other spots where I found "then" missing.

annevk commented on source:
```diff
@@ -5325,6 +5333,274 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+    <ol>
+     <li><p>If <var>element</var>'s <span data-x="explicitly set attr-element">explicitly-set
+     <var>attr</var>-element</span> is a <span>descendant</span> of any of <var>element</var>'s
+     <span data-x="concept-shadow-including-ancestor">shadow-including ancestors</span>, then let
```
No nested let please.

annevk replied:
Why can't we use early returns in this algorithm? Wouldn't that be much more readable?

dandclark replied:
Removed nested let and restructured this algorithm to use early returns.

annevk commented on source:
```diff
@@ -8516,66 +8758,25 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
-  <code>Element</code>, then with <var>attr</var> being the <span>reflected content attribute
-  name</span>:</p>
+  <p>If a <span>reflected IDL attribute</span> has the type <code>Element</code>, then with
+  <var>attr</var> being the <span>reflected content attribute name</span>:</p>
```
Even if we remove inheritance I'm pretty sure we have to handle `?`.

dandclark replied:
I think the `?` was mistakenly dropped. It still needs to be nullable. I've changed it to `Element?`.

annevk commented on source:
```diff
@@ -36247,7 +36394,7 @@ interface <dfn interface>HTMLObjectElement</dfn> : <span>HTMLElement</span> {
   [<span>CEReactions</span>, <span data-x="xattr-Reflect">Reflect</span>] attribute DOMString <dfn attribute for="HTMLObjectElement" data-x="dom-object-type">type</dfn>;
   [<span>CEReactions</span>, <span data-x="xattr-Reflect">Reflect</span>] attribute DOMString <dfn attribute for="HTMLObjectElement" data-x="dom-object-name">name</dfn>;
-  readonly attribute <span>HTMLFormElement</span>? <span data-x="dom-fae-form">form</span>;
+  readonly attribute <span>HTMLElement</span>? <span data-x="dom-fae-form">form</span>;
```
How does this work if we say it has to be `Element` above? (Not the only occurrence it seems.)

dandclark replied:
I think it was a mistake for the type to be `Element` above; I've changed it to `Element?`. So that should fit with the usage here and elsewhere.

annevk replied:
I think you misunderstood my comment here. This was specifically talking about the mismatch between `Element` and `HTMLElement`.

dandclark replied:
Ah, sorry, I understand now.

This (and the other IDL `form` attributes like it) isn't defined as a reflecting attribute, so it doesn't depend on the "If a reflected IDL attribute has the type `Element?`" algorithm, so the type doesn't need to match with that. This has its own definition from further down in the spec, which this PR also modifies to be:

> Listed form-associated elements except for form-associated custom elements have a form IDL attribute, which, on getting, must return the result of retargeting the element's form owner against the element, or null if there isn't a form owner.

I think in practice this would always end up being an `HTMLElement`, since it will either be the actual `HTMLFormElement` or it'll be a shadow host due to retargeting, and currently only `HTMLElement` can be a shadow host due to the namespace check in step 1 of [attach a shadow root](https://dom.spec.whatwg.org/#concept-attach-a-shadow-root).

But on the other hand:
- It's confusing for this to be `HTMLElement` since the related algorithms all deal in `Element`s
- Non-`HTMLElement`s can be shadow hosts in Blink due to an [apparently-longstanding bug](https://issues.chromium.org/issues/521896365)
- Maybe someday we'll want to deliberately [allow certain non-HTML-elements to be shadow hosts](https://github.com/WICG/webcomponents/issues/634).

So all-in-all you're probably right that these should just be `Element`. I've updated all the occurrences.

annevk commented on source:
```diff
@@ -85125,6 +85305,9 @@ dictionary <dfn dictionary>ToggleEventInit</dfn> : <span>EventInit</span> {
   slot (or map of additional fields) is required to properly specify this.</p>
 
+  <p>The <code data-x="dom-ToggleEvent-source">source</code> attribute defines the <span
+  data-x="concept-event-source">source</span> for a <code>ToggleEvent</code>.</p>
```
This is too much action-at-a-distance. We need to actually define this at dispatch time.

dandclark replied:
Acknowledged. I'll work on this.

annevk commented on source:
```diff
@@ -145643,6 +145832,19 @@ document.body.appendChild(text);
        data-x=""> shadowrootcustomelementregistry=&quot;&quot;</code>".</p></li>
 
+       <li>
+        <p>If <var>shadow</var>'s <span>reference target</span> is non-null:</p>
```
Is there a test for the serialization order? Why is this after registry when elsewhere you always put it before the registry?

dandclark replied:
I've updated the ordering here to match the order in the DOM spec PR (reference target before registry).

I'd missed including a test for this. I'm adding it here: https://github.com/web-platform-tests/wpt/pull/60213

dandclark committed:
Address most of Anne's feedback

dandclark committed:
Align getHTML serialization order with ordering in DOM spec (before custom element registry)

dandclark committed:
Use source and target instead of X and Y

dandclark committed:
Reword that single-element reference is specified as content attribute value

dandclark commented:
@annevk I think this basically makes sense. There are a couple details I'm unclear on about setting the event to composed.
- It seems like there is some additional compatibility risk to this. For example, consider the case when the `submit` event is fired on a form due to a button being clicked, with no reference target involved, so the button and form are in the same tree. There would be no change to how the event bubbles, since both `target` (the form) and `relatedTarget` (the button) are in the same tree, but event listeners on the form would still see `event.composed == true`.
  - Maybe we can avoid this by only setting composed if `target` and `relatedTarget` are not in the same tree, which would mean that reference target was being used, which removes the compat concern since it's a new thing?
- When you say "we set composed to true at the time we pass the event to the dispatch algorithm", when exactly does this mean? Usually composed gets set from the HTML spec when it calls [fire an event](https://dom.spec.whatwg.org/#concept-event-fire), so maybe we could do it at that time when the HTML spec calls [fire an event](https://dom.spec.whatwg.org/#concept-event-fire) for Command, Toggle, and Submit events. But it sounds like your idea might have been that we'd set composed inside the DOM spec algos either at the beginning of [dispatch](https://dom.spec.whatwg.org/#concept-event-dispatch) or right before dispatch, in [fire an event](https://dom.spec.whatwg.org/#concept-event-fire)? This might be OK but seems like it could cause compat issues for events like mouseenter that already use `relatedTarget` but are not composed; we’d have to exclude them from the change somehow like by scoping the change to specific event types.

dandclark committed:
Fix link to event target, add link to relatedTarget concept, change fire a focus event to pass relatedTarget to be set as concept rather than attribute

dandclark committed:
Remove event source concept. When submit, toggle/beforetoggle, and command are fired, set composed and relatedTarget.

dandclark committed:
Define CommandEvent and ToggleEvent's source as returning relatedTarget

dandclark committed:
Define SubmitEvent's submitter in terms of relatedTarget

annevk commented on source:
```diff
@@ -8908,69 +9172,18 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
-  where <var>T</var> is either <code>Element</code> or an interface that inherits from
-  <code>Element</code>, then with <var>attr</var> being the <span>reflected content attribute
-  name</span>:</p>
+  <p>If a <span>reflected IDL attribute</span> has the type <code>Element</code>?, then with
```
This should be `<code data-x="">Element?</code>` I think. Or you reuse the original phrasing but restrict T to Element alone.

dandclark replied:
`<code data-x=""><span>Element</span>?</code>` keeps the link to the Element IDL with the right syntax highlighting and is consistent with "If a reflected IDL attribute has the type `DOMString?`", so I changed it to that.

annevk commented on source:
```diff
@@ -85114,10 +85284,8 @@ dictionary <dfn dictionary>ToggleEventInit</dfn> : <span>EventInit</span> {
-  data-x="dom-ToggleEvent-source">source</code> against <span>this</span>'s <code
-  data-x="dom-Event-currentTarget">currentTarget</code>.</p>
+  data-x="dom-ToggleEvent-source">source</code></dfn> getter steps are to return
+  <span>this</span>'s <span data-x="concept-event-relatedTarget">relatedTarget</span>.</p>
```
How does this work for synthetic events. Should we remove the `source` member from IDL or do we allow an explicitly-set `source` to win somehow? (Same for the other events.) Needs test coverage as well.

dandclark replied:
The simplest thing could be that when `source` is set on a synthetic event (`new CommandEvent('command', {source: someElement}` etc), it sets the actual Event relatedTarget concept, just like with non-synthetic events that have a source. So it could affect the event path, and could be retargeted during dispatch.

The downside of this would be compatability. Devs can already create synthetic events today without the use of reference target whose dispatch behavior would be affected by this. Are there other reasons not to prefer that approach?

If that sounds bad due to the compat issue or for other reasons, seems reasonable to have something like an explicitly-set `source` that doesn't touch `relatedTarget` but that could be reflected in the IDL attribute.

dandclark replied:
Looking at this again, there's an error in my response above. Based on the current spec text, `composed` wouldn't get set for synthetic events even if `source` is given, because the current spec only sets it only in spots where trusted events are getting created via `fire an event`. Tests added at https://wpt.fyi/results/shadow-dom/reference-target/tentative/event-path.html reflect this.

So for cases like `new CommandEvent("command", {source: button})`, the event dispatch wouldn't be affected even though the internal *relatedTarget* is set to `button`, since `composed` isn't set.

The one interesting case is `new CommandEvent("command", {source: button, composed: true})`. Here the internal *relatedTarget* is set to `button`, and since `composed` is manually set to true, the event can dispatch in a parent shadow if `button` is in a parent shadow. I think the compat risk of that change is minimal though.

annevk replied:
I don't think we want synthetic event construction to be able to set the internal relatedTarget concept. At least that seems a bit magical to me. @smaug----?

smaug---- replied:
I would expect composed to be initialized to whatever is passed to constructor, and .source would initialize the internal relatedTarget concept, just like what browsers do with FocusEvent and its relatedTarget

dandclark replied:
Initializing the internal relatedTarget concept when a dev does `new CommandEvent("command", {source: button})` etc sounds reasonable to me. I'm unsure about what's the best way to do this in the spec though.
It "just works" with the `FocusEvent` constructor setting `relatedTarget` since it's set from the `relatedTarget` property of the init dictionary, but here we need to map `source` --> `relatedTarget` somehow, and as https://github.com/whatwg/dom/issues/1328 discusses there isn't really a way to do this today.

It doesn't seem right to block this PR on https://github.com/whatwg/dom/issues/1328 being solved. In the meantime, what's the best way to do this? I'm thinking that maybe the HTML spec can define constructor steps for CommandEvent/ToggleEvent/SubmitEvent that set the internal relatedTarget and then call out to the DOM spec for the rest. So mirroring https://dom.spec.whatwg.org/#concept-event-constructor, that would consist of adding this under https://html.spec.whatwg.org/#the-commandevent-interface (and the same for ToggleEvent/SubmitEvent):

The `new CommandEvent(type, eventInitDict)` constructor steps are:
1. Let `event` be the result of running the [inner event creation steps](https://dom.spec.whatwg.org/#inner-event-creation-steps) with `CommandEvent`, `null`, `now`, and `eventInitDict`.
2. Initialize `event`’s `type` attribute to `type`.
3. If `eventInitDict[source]` exists, set `event`'s `relatedTarget` to `eventInitDict[source]`.
4. Return `event`.

Does that seem reasonable?

dandclark replied:
I realized this can be done more simply by defining [event constructing steps](https://dom.spec.whatwg.org/#concept-event-constructor-ext) for each of these events to map the initDict member to the internal relatedTarget. I've pushed that change (https://github.com/whatwg/html/pull/10995/changes/8adf3afed69aab39e47f7d36400993be21cf38e8).

smaug---- commented on source:
```diff
@@ -66228,14 +66390,18 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+   data-x="dom-ToggleEvent-newState">newState</code> attribute initialized to "<code
+   data-x="">open</code>", the <code data-x="dom-ToggleEvent-source">source</code> attribute
+   initialized to <var>source</var>, the <span>composed flag</span> set to true, and given <span
+   data-x="concept-event-relatedTarget">relatedTarget</span> <var>source</var>.</p></li>
```
I thought we wanted composed to be false if source is null, with toggle events and.. command and what else? submit?

dandclark replied:
Oops, that's right. Fixed. It's for `toggle`/`beforetoggle`, `command`, and `submit` events.

dandclark committed:
Only set composed if source/submitter is non null. Use link for relatedTarget parameter

dandclark committed:
Also use linked named param for legacy target override flag

dandclark committed:
Put the ? back inside the <code> element

dandclark committed:
Change HTMLElement getters to Element

annevk committed:
AI-found blocking issues (leaving Event getter issue unfixed for now)

annevk committed:
AI nits

annevk committed:
fix more issues

dandclark committed:
Add event constructing steps for SubmitEvent, ToggleEvent, and CommandEvent to set relatedTarget concept to submitter/source passed in event constructor init dict.

dandclark committed:
Update reference to "get the resolved attribute target element" algo in "get the target popover".

annevk commented:
@dandclark I had AI perform a check and it still identifies many structural issues with this PR (and it appears to be correct). See https://gist.github.com/annevk/48545a644f86e2536bafc5c437a64517 (starting at line 44 for this PR, the DOM PR is above).
